The goal here is to use nix to manage dependencies so that we get a reproducible derivation every time.
One simple way of doing that would be to look at nix-packages for `cc-test-reporter` but since its not there we can write a simple derivation to get `cc-test-reporter` from github release page.

Here is how:

The release url can be fetched from here -> https://github.com/codeclimate/test-reporter?tab=readme-ov-file#download
Lets see how a simple nix derivation for fetching a binary of `cc-test-reporter` would look like

In nix we use the famous `mkDerivation` function to make a derivation and it is present inside `stdenv`
Inside `mkDerivation` we can fetch a binary from any url using `fetchurl` function which has to be imported separately.

```nix

{ stdenv, fetchurl }:

stdenv.mkDerivation rec {
  pname = "cc-test-reporter";

  src = fetchurl {
    url = "https://codeclimate.com/downloads/test-reporter/test-reporter-0.11.1-darwin-amd64";
    hash = "sha256-somerandomvaluefornow";
   };

  dontUnpack = true;

  installPhase = ''
    install -D $src $out/bin/cc-test-reporter
    chmod +x $out/bin/cc-test-reporter
  '';

}

```

some explanation of the code snippet above : 
- `{ stdenv, fetchurl }:` -> This is the input/import section of the nix derivation.
- `dontUnpack = true;` -> we do this because when we fetch the URL we don't want to unpack it ( since its not a tar ball)
- `hash = "sha256-somerandomvaluefornow";` -> we will generate hash in the next step so ignore this for now.

now how do we test if this works?
-> We have to add this nix derivation in pkgs directory.
-> We have to add an entry in overlay.nix to call this package.
-> We have to add this derivation to shell.nix so that we can do `make shell` and see if we are able to access `cc-test-reporter` or not.

Lets go.

- Make a file named `default.nix` here -> `status-go/nix/pkgs/test-reporter/default.nix` and paste the code from the snippet shared above.

- Then add the reference to this derivation in our `overlay.nix` like this
```diff
diff --git a/nix/overlay.nix b/nix/overlay.nix
index 4b528356d..eb24c69e2 100644
--- a/nix/overlay.nix
+++ b/nix/overlay.nix
@@ -51,6 +51,7 @@ in rec {

   # Custom packages
   go-modvendor = callPackage ./pkgs/go-modvendor { };
+  cc-test-reporter = callPackage ./pkgs/test-reporter { };

   gomobile = (prev.gomobile.overrideAttrs (old: {
     patches = [
```
- Then add `cc-test-reporter` to shell like this 

```diff
diff --git a/nix/shell.nix b/nix/shell.nix
index b8f8e2697..cb3b1d469 100644
--- a/nix/shell.nix
+++ b/nix/shell.nix
@@ -21,7 +21,7 @@ in pkgs.mkShell {
   buildInputs = with pkgs; [
     git jq which
     go golangci-lint go-junit-report gopls go-bindata gomobileMod
-    mockgen protobuf3_20 protoc-gen-go gotestsum go-modvendor openjdk
+    mockgen protobuf3_20 protoc-gen-go gotestsum go-modvendor openjdk cc-test-reporter
    ] ++ lib.optionals (stdenv.isDarwin) [ xcodeWrapper ];

    shellHook = lib.optionalString (!isMacM1) '
```

now execute `make shell` in terminal and you should see an error like this one : 

```
error: hash mismatch in fixed-output derivation '/nix/store/k8clyncy6i0jigzgpvkh3gmvap50vgxg-test-reporter-0.11.1-darwin-amd64.drv':
         specified: sha256-somerandomvaluefornow
            got:    sha256-uO9aRL3cJe+KCoC+uN6cBQy8xGQHim6h5Qzw36QO7EY=
error: 1 dependencies of derivation '/nix/store/2yd2h5ppk8yivqivycmwfmcdr10p20iz-cc-test-reporter-0.11.1.drv' failed to build
make: *** [shell] Error 1
```
This is nix telling you the correct hash for the binary we want to download. lets use this value in our derivation. 
At `status-go/nix/pkgs/test-reporter/default.nix` update the snippet to match the hash like this : 

```nix 
  src = fetchurl {
    url = "https://codeclimate.com/downloads/test-reporter/test-reporter-0.11.1-darwin-amd64";
    hash = "sha256-uO9aRL3cJe+KCoC+uN6cBQy8xGQHim6h5Qzw36QO7EY=";
   };
```
and then exit shell and re-run `make shell` should build the derivation properly and we would now have `cc-test-reporter` available in our nix shell for usage :)

note : This is a very basic derivation and will work only for macOS Environments, you could use your imagination (or look at existing derivations) and make sure it builds for Linux,Windows as well (if needed).

ps: I could have just pushed a commit with a working derivation but a wise man once said : 
> “If you give a man a fish, you feed him for a day. If you teach a man to fish, you feed him for a lifetime.”

Happy Fishing @igor-sirotin

--- a/Makefile
+++ b/Makefile
@@ -351,6 +351,7 @@ test-unit: export UNIT_TEST_COUNT ?= 1
 test-unit: export UNIT_TEST_FAILFAST ?= true
 test-unit: export UNIT_TEST_RERUN_FAILS ?= true
 test-unit: export UNIT_TEST_USE_DEVELOPMENT_LOGGER ?= true
+test-unit: export UNIT_TEST_REPORT_CODECLIMATE ?= false
 test-unit: export UNIT_TEST_PACKAGES ?= $(call sh, go list ./... | \
 	grep -v /vendor | \
 	grep -v /t/e2e | \
@@ -366,7 +367,13 @@ test-unit: export UNIT_TEST_PACKAGES_NOT_PARALLELIZABLE ?= \
 test-unit: export UNIT_TEST_PACKAGES_WITH_EXTENDED_TIMEOUT ?= \
 	github.com/status-im/status-go/protocol
 test-unit: ##@tests Run unit and integration tests
+ifdef UNIT_TEST_REPORT_CODECLIMATE
+	cc-test-reporter before-build
+endif
 	./_assets/scripts/run_unit_tests.sh
+ifdef UNIT_TEST_REPORT_CODECLIMATE
+	cc-test-reporter after-build --prefix=github.com/status-im/status-go
+endif
 
 test-unit-race: export GOTEST_EXTRAFLAGS=-race
 test-unit-race: test-unit ##@tests Run unit and integration tests with -race flag

--- a/_assets/ci/Jenkinsfile.tests
+++ b/_assets/ci/Jenkinsfile.tests
@@ -30,6 +30,11 @@ pipeline {
       defaultValue: getDefaultUnitTestUseDevelopmentLogger(),
       description: 'Should the job use detailed logging for tests, potentially generating large logs?'
     )
+    booleanParam(
+      name: 'UNIT_TEST_REPORT_CODECLIMATE',
+      defaultValue: getDefaultUnitTestReportCodeClimate(),
+      description: 'Should the job report test coverage to CodeClimate?'
+    )
   }
 
   options {
@@ -120,15 +125,13 @@ pipeline {
           "--publish=${DB_PORT}:${DB_PORT}",
         ].join(' '), "-p ${DB_PORT}") { c ->
           nix.shell('make generate-handlers', pure: true)
-          sh "./cc-test-reporter before-build"
-          nix.shell('make test-unit V=1', pure: false)
           withCredentials([
             string(
               credentialsId: 'codeclimate-test-reporter-id',
               variable: 'CC_TEST_REPORTER_ID'
             ),
           ]) {
-            sh "./cc-test-reporter after-build --prefix=github.com/status-im/status-go"
+            nix.shell('make test-unit V=1', pure: false)
           }
         }
       } }
@@ -185,6 +188,8 @@ def getDefaultUnitTestCount() { isTestNightlyJob() ? '20' : '1' }
 def getDefaultUnitTestRerunFails() { isTestNightlyJob() ? false : true }
 
 def getDefaultUnitTestUseDevelopmentLogger() { isTestNightlyJob() ? false : true }
+
+def getDefaultUnitTestReportCodeClimate() { isTestNightlyJob() ? false : true }
 
 def getDefaultUnitTestFailfast() { isTestNightlyJob() ? false : true }
 

--- a/_assets/ci/Jenkinsfile.tests
+++ b/_assets/ci/Jenkinsfile.tests
@@ -107,13 +107,6 @@ pipeline {
       } }
     }
 
-    stage('Setup CodeClimate') {
-      steps { script {
-        sh "curl -L https://codeclimate.com/downloads/test-reporter/test-reporter-latest-linux-amd64 > ./cc-test-reporter"
-        sh "chmod +x ./cc-test-reporter"
-      } }
-    }
-
     stage('Unit Tests') {
       environment {
         TEST_POSTGRES_PORT = "${env.DB_PORT}"

--- a/_assets/ci/Jenkinsfile.tests
+++ b/_assets/ci/Jenkinsfile.tests
@@ -102,7 +102,7 @@ pipeline {
       } }
     }
 
-    stage('Setup Codeclimate') {
+    stage('Setup CodeClimate') {
       steps { script {
         sh "curl -L https://codeclimate.com/downloads/test-reporter/test-reporter-latest-linux-amd64 > ./cc-test-reporter"
         sh "chmod +x ./cc-test-reporter"
@@ -122,7 +122,14 @@ pipeline {
           nix.shell('make generate-handlers', pure: true)
           sh "./cc-test-reporter before-build"
           nix.shell('make test-unit V=1', pure: false)
-          sh "./cc-test-reporter after-build --prefix=github.com/status-im/status-go"
+          withCredentials([
+            string(
+              credentialsId: 'codeclimate-test-reporter-id',
+              variable: 'CC_TEST_REPORTER_ID'
+            ),
+          ]) {
+            sh "./cc-test-reporter after-build --prefix=github.com/status-im/status-go"
+          }
         }
       } }
       post { cleanup { /* Leftover DB containers. */

--- a/nix/overlay.nix
+++ b/nix/overlay.nix
@@ -51,6 +51,7 @@ in rec {
 
   # Custom packages
   go-modvendor = callPackage ./pkgs/go-modvendor { };
+  cc-test-reporter = callPackage ./pkgs/cc-test-reporter { };
 
   gomobile = (prev.gomobile.overrideAttrs (old: {
     patches = [

--- a/nix/pkgs/cc-test-reporter/default.nix
+++ b/nix/pkgs/cc-test-reporter/default.nix
@@ -1,0 +1,31 @@
+{ lib, stdenv, fetchurl }:
+
+let
+  inherit (stdenv) isLinux isDarwin isWindows;
+
+  platform =
+    if isLinux then "linux" else
+    if isDarwin then "darwin" else
+    if isWindows then "windows" else
+    throw "Unsupported platform: ${stdenv.hostPlatform.system}";
+
+in stdenv.mkDerivation rec {
+  pname = "cc-test-reporter";
+  version = "0.11.1";
+
+  src = fetchurl {
+    url = "https://codeclimate.com/downloads/test-reporter/test-reporter-${version}-${platform}-amd64";
+    hash = lib.getAttr platform {
+      darwin = "sha256-uO9aRL3cJe+KCoC+uN6cBQy8xGQHim6h5Qzw36QO7EY=";
+      linux = "ne79mW3w9tHJ+3lAWzluuRp6yjWsy4lpdV/KpmjaTa0=";
+      windows = "8pn8csW9l5xMerZWAwIwWcrO7OLNWEM03yPEMMllaak=";
+    };
+   };
+
+  dontUnpack = true;
+
+  installPhase = ''
+    install -D $src $out/bin/cc-test-reporter
+    chmod +x $out/bin/cc-test-reporter
+  '';
+}

--- a/nix/shell.nix
+++ b/nix/shell.nix
@@ -21,7 +21,7 @@ in pkgs.mkShell {
   buildInputs = with pkgs; [
     git jq which
     go golangci-lint go-junit-report gopls go-bindata gomobileMod
-    mockgen protobuf3_20 protoc-gen-go gotestsum go-modvendor openjdk
+    mockgen protobuf3_20 protoc-gen-go gotestsum go-modvendor openjdk cc-test-reporter
    ] ++ lib.optionals (stdenv.isDarwin) [ xcodeWrapper ];
 
    shellHook = lib.optionalString (!isMacM1) ''

